### DISCUSSION TITLE
two-step login with yubi does not seem to work with waf

### DISCUSSION THREAD

**Discussion by eleaner** [dani-garcia/vaultwarden#3042] [Q&A]
### Subject of the issue
It is probably rather missing element in recipe than in the vaultwarden itself
however it would then be an error in WiKi

after enabling two-step login an attempt to login through hardened docker installation ends with an email like below

log in directly to the vaulwardedn container (exposed port 80) works perfectly

![image](https://user-images.githubusercontent.com/5169203/209483423-bb85cc32-350d-4ba1-8587-c0b8dcf53ec4.png)

### Deployment environment

* vaultwarden version: 2022.10.0

* Install method:docker-compose following this:
https://github.com/dani-garcia/vaultwarden/wiki/Docker---Traefik---ModSecurity-Setup

* Clients used: web-ui

* Reverse proxy and version: traefik: latest

* MySQL/MariaDB or PostgreSQL version: n/a

* Other relevant details:

### Steps to reproduce

following this recipe
https://github.com/dani-garcia/vaultwarden/wiki/Docker---Traefik---ModSecurity-Setup
enable yubikey authentication with ybikey

### Expected behaviour

I am receiving a prompt to authenticate
![image](https://user-images.githubusercontent.com/5169203/209483336-0e330f0c-6dd9-4a1a-abdf-618f6fe011e8.png)

### Actual behaviour

I am receiving an error and email mentioned above
![image](https://user-images.githubusercontent.com/5169203/209483352-58374b71-314f-473c-a7a4-672d31bc242a.png)

**Comment by BlackDex** [comment]
We would need logs to know anything.
Did the request went to Vaultwarden at all, did the reverse proxy received it? So at least since logs of either (or both) are needed to be able to help and say something useful.

**Comment by eleaner** [comment]
here the log (I tried to obscure some of the info for security - but not even sure if the correct one)

```
vaultwarden    | [2022-12-25 22:44:38.551][request][INFO] POST /api/accounts/prelogin
vaultwarden    | [2022-12-25 22:44:38.555][response][INFO] (prelogin) POST /api/accounts/prelogin => 200 OK
vaultwarden    | [2022-12-25 22:44:38.655][request][INFO] POST /identity/connect/token
vaultwarden    | [2022-12-25 22:44:39.198][error][ERROR] 2FA token not provided
vaultwarden    | [2022-12-25 22:44:39.198][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
waf            | {"transaction":{"time":"25/Dec/2022:22:44:39.200564 +0000","transaction_id":"Y6jSVrjIh9XLjmbolYfYgQAAAEk","remote_address":"xxx.xxx.xxx.xxx","remote_port":37592,"local_address":"172.20.0.4","local_port":80},"request":{"request_line":"POST /identity/connect/token HTTP/1.1","headers":{"Host":"vw.zone667.com","User-Agent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36","Content-Length":"231","Accept":"application/json","Accept-Encoding":"gzip, deflate, br","Accept-Language":"en-GB,en-US;q=0.9,en;q=0.8","Auth-Email":"dndAem9uZTY2Ny5jb20","Bitwarden-Client-Name":"web","Bitwarden-Client-Version":"2022.10.0","Content-Type":"application/x-www-form-urlencoded; charset=utf-8","Device-Type":"9","Origin":"https://vw.zone667.com","Referer":"https://vw.zone667.com/","Sec-Ch-Ua":"\"Not?A_Brand\";v=\"8\", \"Chromium\";v=\"108\", \"Brave\";v=\"108\"","Sec-Ch-Ua-Mobile":"?0","Sec-Ch-Ua-Platform":"\"macOS\"","Sec-Fetch-Dest":"empty","Sec-Fetch-Mode":"cors","Sec-Fetch-Site":"same-origin","Sec-Gpc":"1","X-Forwarded-Host":"vw.zone667.com","X-Forwarded-Port":"443","X-Forwarded-Proto":"https","X-Forwarded-Server":"1b801915d956","X-Real-Ip":"86.10.58.250"},"body":["scope=api%20offline_access&client_id=web&deviceType=9&deviceIdentifier=93d685f4-0846-4568-a635-78172f2f5fb3&deviceName=chrome&grant_type=password&username=vw%40zone667.com&password=xxxxxxx%2xxxxxxxx"]},"response":{"protocol":"HTTP/1.1","status":400,"headers":{"Content-Length":"226","Connection":"close","Content-Type":"text/html; charset=iso-8859-1"},"body":"<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n<html><head>\n<title>400 Bad Request</title>\n</head><body>\n<h1>Bad Request</h1>\n<p>Your browser sent a request that this server could not understand.<br />\n</p>\n</body></html>\n"},"audit_data":{"handler":"proxy-server","stopwatch":{"p1":581,"p2":3022,"p3":166,"p4":602,"p5":127,"sr":110,"sw":1,"l":0,"gc":0},"response_body_dechunked":true,"producer":["ModSecurity for Apache/2.9.6 (http://www.modsecurity.org/)","OWASP_CRS/3.3.4"],"server":"Apache","engine_mode":"ENABLED"}}
```

**Reply by BlackDex** [reply]
Looks like a ModSecurity issue to me. It seems to block something for some reason. Check the logs of Apache or try to enable debug logging for it and check again.

Because it doesn't look like this request went to Vaultwarden at all.

**Reply by eleaner** [reply]
still, one cannot fully run the Vaultwarden following the recipe in WiKi :(
shame

**Reply by BlackDex** [reply]
If you have any improvements, you can add it. The wiki is maintained by everybody. It's not an official documentation.
